### PR TITLE
[dm][cli] Spring 15 and 16 stories

### PR DESCRIPTION
**Pull request by kopachevsky**
- Add flag to print dependency graph
    - Implement Delete action
    - Implement Apply action
    - Implement Update action
    - Add output formatting for deployment creating Stages same way existing CLI does
    - Cross-deployment references for complex yaml (not just plain string)
    - Implement advanced options to pass config yamls (strings, directory, pattern)
    - Implement default project id setup according to configuration rules
    - Implement preview mode for update/apply action
    - Set deployment name from file name in case it missing on config yaml
    - Added integration testing

### CODE REVIEW

File: cli/deployment/api_client_test.go [hunk 1]
```diff
@@ -24,3 +24,27 @@ func TestGetOutputs(t *testing.T) {
 		t.Errorf("expected: %s, got: %s", expected, outputs["my-network-prod.name"])
 	}
 }
+
+/*
```
Comment (morgante) [inline cli/deployment/api_client_test.go]: @kopachevsky What's the purpose of this commented-out code? It shouldn't be included.

Comment (kopachevsky) [reply cli/deployment/api_client_test.go]: @morgante this test uses real API calls, that why I "ignored" it with comments, will remote this

File: cli/deployment/api_client.go [hunk 2]
```diff
@@ -11,73 +11,233 @@ import (
...
+		log.Printf("failed to start cmd: %v, output: %v", err, string(output))
+		return string(output), err
 	}
 
-	return outBuff.String(), err
+	return string(output), err
 }
 
 // GetOutputs retrive existing Deployment outputs using gcloud and store result in map[string]string
 // where "resourceName.propertyName" is key, and value is string representation of the output value.
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: The type returned by this function has changed, so please update the documentation comment for this function.

File: cli/deployment/api_client.go [hunk 3]
```diff
@@ -11,73 +11,233 @@ import (
...
 		log.Printf("Failed to get deployment manifest: %v", err)
 		return nil, err
 	}
 	return parseOutputs(data)
 }
 
-// Create creates a new Deployment based on a Deployment object passed into it.
-func Create(deployment *Deployment) (*Deployment, error) {
+/*
+returns project id taken from local gcloud configuration
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: Per https://github.com/golang/go/wiki/CodeReviewComments#doc-comments, this function should have a documentation comment. Prefer using line (`//`-style) comments for this sort of documentation rather than block comments.

File: cli/deployment/api_client.go [hunk 4]
```diff
@@ -11,73 +11,233 @@ import (
...
+		}
+	}{}
+	err = yaml.Unmarshal([]byte(data), &out)
+	if err != nil {
+		return "", err
+	}
+	return out.Core.Project, nil
+}
+
+// Create deployment based on passed Deployment object passed into it.
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: Per https://golang.org/doc/effective_go.html#commentary, doc comments for exported functions should generally be of the form "CreateOrUpdate creates a new deployment based on the Deployment object passed into it..." Since the function name has changed, this sentence should be changed to match.

Since the function creates or updates based on what the user has supplied, this might get confusing. I'd recommend making separate `Create` and `Update` functions that call this function, and making this one unexported. Then the doc comments for Create and Update are clear and the API is cleaner.

File: cli/deployment/api_client.go [hunk 5]
```diff
@@ -11,73 +11,233 @@ import (
...
+			log.Printf("on %s action, failed to get outputs for deployment: %v, error: %v", action, deployment, err)
+			return output, err
+		}
+		deployment.Outputs = outputs
+	}
+	return output, nil
+}
+
+/*
+cancel update/create/delete preview with gcloud deployments cancel-preview command,
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: Please use line comments as suggested previously and follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/api_client.go [hunk 6]
```diff
@@ -11,73 +11,233 @@ import (
...
+	}
+	output, err := runGCloud(args...)
+	if err != nil {
+		log.Printf("failed to get deployment manifest: %v", err)
+		return output, err
+	}
+	return output, nil
+}
+
+func GetStatus(deployment *Deployment) (Status, error) {
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: Please add a doc comment for this exported function as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/config.go [hunk 7]
```diff
@@ -3,12 +3,16 @@ package deployment
...
 	"log"
+	"os"
+	"path/filepath"
 	"regexp"
 	"strings"
 
 	"gopkg.in/yaml.v2"
 )
 
+var DefaultProjectID string
```
Comment (dhobsd) [inline cli/deployment/config.go]: Please add a doc comment for this exported variable, or make it unexported.

File: cli/deployment/config.go [hunk 8]
```diff
@@ -18,6 +22,7 @@ type Config struct {
 	// config YAML field and if it missing in YAML, from config YAML file name itself.
 	Name string
 	// Project field contains GCP Project Id it could be initialized from config YAML, env variable, gcloud default.
+	// Node: don't use this varible directly, use GetProjectID instead!!!
```
Comment (dhobsd) [inline cli/deployment/config.go]:
```suggestion
	// Note: don't access this member directly, use GetProjectID instead.
```

File: cli/deployment/config.go [hunk 9]
```diff
@@ -41,17 +48,59 @@ func NewConfig(data string, file string) Config {
...
+// GetProject returns project id if it set in config file, otherwise default value defined in DefaultProjectID var.
+func (c Config) GetProject() string {
+	if len(c.Project) > 0 {
+		return c.Project
+	} else {
+		return DefaultProjectID
+	}
+}
+
+// returns file path in case of file source or yaml string in case of yaml source
```
Comment (dhobsd) [inline cli/deployment/config.go]: Per https://golang.org/doc/effective_go.html#commentary this should read:

```suggestion
// Source returns the the path of the file for a file-backed configuration, or a YAML string for YAML configurations.
```

File: cli/deployment/config.go [hunk 10]
```diff
@@ -61,15 +110,19 @@ func (c Config) String() string {
 
 // The YAML function marshals a Config object to YAML. It sets all relative import paths to absolute paths and removes
 // all custom elements that the gcloud deployment manager is not aware of, including name, project, and description.
-func (c Config) YAML() ([]byte, error) {
+func (c Config) YAML(outputs map[string]map[string]interface{}) ([]byte, error) {
```
Comment (dhobsd) [inline cli/deployment/config.go]: Please update the doc comment to explain this new argument to the function. This would've made this review easier; I had to look through a lot of files to finally figure out that in fact this is an in/out parameter. Please also rename the argument to better reflect its purpose; as an in/out parameter, a name like "outputs" is misleading.

File: cli/deployment/config.go [hunk 11]
```diff
@@ -129,7 +181,70 @@ func (c Config) findAllOutRefs() []string {
...
 
+func getOutRefValue(ref string, outputs map[string]map[string]interface{}) interface{} {
+	fullName, res, name := parseOutRef(ref)
+	outputsMap := outputs[fullName]
+	if outputsMap == nil {
+		arr := strings.Split(fullName, ".")
+		var err error
+		outputsMap, err = GetOutputs(arr[0], arr[1])
+		if err != nil {
+			log.Fatalf("Erorr getting outputs for deployment: %s, error: %v", fullName, err)
```
Comment (dhobsd) [inline cli/deployment/config.go]:
```suggestion
			log.Fatalf("Error getting outputs for deployment: %s, error: %v", fullName, err)
```

File: cli/deployment/config.go [hunk 12]
```diff
@@ -129,7 +181,70 @@ func (c Config) findAllOutRefs() []string {
...
+		}
+		outputs[fullName] = outputsMap
+	}
+	value, ok := outputsMap[res+"."+name]
+	fullRef := fmt.Sprintf("$(out.%s)", ref)
+	if !ok {
+		log.Fatalf("Unresolved dependency: %s. Deployment: %s , on whichother resources depended, neither was specifiedin the submitted congigs nor existed inDeployment Manager", fullRef, fullName)
```
Comment (dhobsd) [inline cli/deployment/config.go]:
```suggestion
		log.Fatalf("Unresolved dependency: %s. Deployment: %s , on which other resources depended, was neither specified in the submitted configs nor existed in Deployment Manager", fullRef, fullName)
```

File: cli/deployment/config.go [hunk 13]
```diff
@@ -41,17 +48,59 @@ func NewConfig(data string, file string) Config {
 		data: data,
 	}
 
+	if len(file) > 0 {
+		config.file = filepath.Clean(file)
+		config.dir = filepath.Dir(file)
+	} else {
+		// yaml passed as sting through parameter
```
Comment (dhobsd) [inline cli/deployment/config.go]:
```suggestion
		// yaml passed as string through parameter
```

File: cli/deployment/config_test.go [hunk 14]
```diff
@@ -86,14 +87,34 @@ func TestFindAllDependencies(t *testing.T) {
...
+
+func TestYamlReplaceOutRefs(t *testing.T) {
+	data := GetTestData("cross-ref", "main-manifest.yaml", t)
+	output, err := parseOutputs(data)
+	if err != nil {
+		t.Errorf("failed to parse outputs: %v", err)
+	}
+	config := NewConfig(GetTestData("cross-ref", "dependent-with-refs.yaml", t), "/user/home/test.yaml")
+	outputs := map[string]map[string]interface{}{}
+	outputs["prj1.name1"] = output
```
Comment (dhobsd) [inline cli/deployment/config_test.go]: This could be combined during assignment:

```
outputs := map[string]map[string]interface{}{"prj1.name1": output}
```

File: cli/deployment/config_test.go [hunk 15]
```diff
@@ -86,14 +87,34 @@ func TestFindAllDependencies(t *testing.T) {
...
+	actual, err := config.YAML(outputs)
+	if err != nil {
+		t.Errorf("failed to export config YAML: %v", err)
+	}
+	ioutil.WriteFile("/1-prj/projects/google/cloud-foundation-toolkit/cli/testdata/cross-ref/test", actual, 0755)
```
Comment (dhobsd) [inline cli/deployment/config_test.go]: This file path probably isn't going to exist for most people who check out this repo. Can you make this a relative path?

File: cli/deployment/config_test.go [hunk 16]
```diff
@@ -86,14 +87,34 @@ func TestFindAllDependencies(t *testing.T) {
...
 	}
 }
+
+func TestYamlReplaceOutRefs(t *testing.T) {
+	data := GetTestData("cross-ref", "main-manifest.yaml", t)
+	output, err := parseOutputs(data)
+	if err != nil {
+		t.Errorf("failed to parse outputs: %v", err)
+	}
+	config := NewConfig(GetTestData("cross-ref", "dependent-with-refs.yaml", t), "/user/home/test.yaml")
```
Comment (dhobsd) [inline cli/deployment/config_test.go]: What is `/user/home/test.yaml`?

File: cli/deployment/deployment.go [hunk 17]
```diff
@@ -22,20 +22,28 @@ type Deployment struct {
 
 // NewDeployment creates a new Deployment object, overriding all outward references.
 // In effect, this means all deployment dependencies must exist in the GCP project.
-func NewDeployment(config Config, outputs map[string]map[string]string) *Deployment {
+func NewDeployment(config Config, outputs map[string]map[string]interface{}, processConfig bool) *Deployment {
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: Please update the comment to describe the new argument.

File: cli/deployment/deployment.go [hunk 18]
```diff
@@ -52,30 +60,43 @@ func (d Deployment) FullName() string {
 	return d.config.FullName()
 }
 
-func replaceOutRefs(config Config, outputs map[string]map[string]string) []byte {
-	data, err := config.YAML()
-	if err != nil {
-		log.Fatalf("error while parsing yaml for config: %s, error: %v", config.FullName(), err)
+func (d *Deployment) Execute(action string, preview bool) (output string, error error) {
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: Please add a doc comment for this exported method as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/deployment.go [hunk 19]
```diff
@@ -52,30 +60,43 @@ func (d Deployment) FullName() string {
...
-			if err != nil {
-				log.Fatalf("Erorr getting outputs for deployment: %s, error: %v", fullName, err)
-			}
-			outputs[fullName] = outputsMap
+
+	if action == ActionCreate || action == ActionUpdate {
+		return CreateOrUpdate(action, d, preview)
+	} else if action == ActionDelete {
+		return Delete(d, preview)
+	} else {
```
Comment (dhobsd) [inline cli/deployment/deployment.go]: This else case is unnecessary. Please remove and outdent

File: cli/deployment/utils.go [hunk 20]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+	baseDir = filepath.Dir(baseDir)
+	relative := filepath.Clean(filepath.Join(baseDir, file))
+	result, err := filepath.Abs(relative)
+	if err != nil {
+		log.Fatalf("error creating absolute path, for file: %s, error: %v", relative, err)
+	}
+	return result
+}
+
+// check if string is yaml
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Please follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary, in particular making sure this doc comment is a complete, grammatical sentence.

File: cli/deployment/utils.go [hunk 21]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+func IsYAML(text string) bool {
+	obj := struct{}{}
+	err := yaml.Unmarshal([]byte(text), obj)
+	return err == nil
+}
+
+// append string map B to map A, returns A
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Please follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary, in particular making sure this doc comment is a complete, grammatical sentence.

It is also worth noting in this comment that any keys in `b` that already exist in `a` are overwritten.

File: cli/deployment/utils.go [hunk 22]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+	return a
+}
+
+/*
+creates valid deployment name from file path satisfied Deployment resource "name" field requirements:
+Specifically, the name must be 1-63 characters long and match the regular expression [a-z]([-a-z0-9]*[a-z0-9])?
+which means the first character must be a lowercase letter,
+and all following characters must be a dash, lowercase letter, or digit, except the last character, which cannot be a dash.
+see more https://cloud.google.com/deployment-manager/docs/reference/latest/deployments#resource
+*/
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Please use line comments as suggested previously and follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/utils.go [hunk 23]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+	firstChar := regexp.MustCompile("^[-0-9]*")
+	lastChar := regexp.MustCompile("-*$")
+	name = firstChar.ReplaceAllString(name, "")
+	name = lastChar.ReplaceAllString(name, "")
+	return name
+}
+
+/*
+As for user input and validate entered value is equal to one of the provided options.
+Returns validated option string
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Please use line comments as suggested previously and follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/utils.go (hunk 22 above)
Comment (dhobsd) [reply cli/deployment/utils.go]: Also note in the documentation that this function truncates the name.

File: cli/deployment/utils.go [hunk 24]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+func DeploymentNameFromFile(path string) string {
+	_, file := filepath.Split(path)
+	name := strings.TrimSuffix(file, filepath.Ext(file))
+	name = strings.ToLower(name)
+	if len(name) > 63 {
+		name = name[0:63]
+	}
+	name = strings.ReplaceAll(name, "_", "-")
+	firstChar := regexp.MustCompile("^[-0-9]*")
+	lastChar := regexp.MustCompile("-*$")
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Why compile this regex every time the function is called?

File: cli/deployment/utils.go [hunk 25]
```diff
@@ -25,13 +30,87 @@ func unmarshal(data string) (map[string]interface{}, error) {
...
+		if err != nil {
+			log.Fatalf("failed to get user input, error: %v", err)
+		}
+	}
+	return input
+}
+
+/*
+Checks if string a in slice
+*/
```
Comment (dhobsd) [inline cli/deployment/utils.go]: Please use line comments as suggested previously and follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary

File: cli/deployment/utils_test.go [hunk 26]
```diff
@@ -42,9 +43,64 @@ func TestReparentPath(t *testing.T) {
 	}
 }
 
+var fileNameTests = []struct {
+	file string
+	name string
+}{
+	{"name", "name"},
```
Comment (dhobsd) [inline cli/deployment/utils_test.go]: Consider also testing the empty string, spaces, single and double quotes, and other odd characters that sometimes appear in filenames.

**Comment by dhobsd** [comment]
Sorry that this took so long. It would be very helpful in the future to have incremental changes submitted in smaller chunks that take less time to review.

File: cli/deployment/utils_test.go (hunk 26 above)
Comment (kopachevsky) [reply cli/deployment/utils_test.go]: done

**Comment by kopachevsky** [comment]
@dhobsd sorry for lots of code, I wasn't able to create incremental review because all next feature was relying on previous, next changes should be much smaller, I've tried to solve all comments, sorry if I missed something

**Comment by ocsig** [comment]
Are you planning to update the documentation together with this PR? ( for example the preview function is not documented in the MD or in the help.

**Comment by kopachevsky** [comment]
> Are you planning to update the documentation together with this PR? ( for example the preview function is not documented in the MD or in the help.

I would update readme in other PR, considering someone native speaker could do this better then myself

**Comment by kopachevsky** [comment]
@dhobsd can you do another round of review? thanks for your efforts!

File: cli/deployment/api_client.go [hunk 27]
```diff
@@ -11,73 +10,245 @@ import (
...
 	if err != nil {
 		log.Printf("Failed to get deployment manifest: %v", err)
 		return nil, err
 	}
 	return parseOutputs(data)
 }
 
-// Create creates a new Deployment based on a Deployment object passed into it.
-func Create(deployment *Deployment) (*Deployment, error) {
+// returns project id taken from local gcloud configuration
```
Comment (dhobsd) [inline cli/deployment/api_client.go]:
```suggestion
// GCloudDefaultProjectID returns the default project id taken from local gcloud configuration.
```

File: cli/deployment/api_client.go [hunk 28]
```diff
@@ -11,73 +10,245 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
+// Status represent Deployment status in enum/numerical format
+type Status int
+
+const (
+	Done     Status = 0
```
Comment (dhobsd) [inline cli/deployment/api_client.go]: Sorry for not catching all of these. Please add doc comments for exported `const` values as well.

**Comment by kopachevsky** [comment]
Apologies! I oversimplified when I said return, but I think you can still restructure this to outdent the normal flow from the error flow:

```
for ... {
  stat, err := os.Stat(arg)
  if os.IsNotExist(err) {
    // os.IsNotExist branch logic
  } else if err != nil {
    log.Fatalf(...)
  }

  // regular, non-error logic
}

return resFiles, resYamls
````
I don't feel this option will work correctly, because os.IsNotExist(err)  assumes error is not nil

**Comment by kopachevsky** [comment]
@dhobsd Devon, please check the fixes, thanks!

**Comment by dhobsd** [comment]
That's not the case; os.IsNotExist/IsExist/IsPermission/IsTimeout all properly handle a `nil` error: https://github.com/golang/go/issues/31065

**Comment by kopachevsky** [comment]
> That's not the case; os.IsNotExist/IsExist/IsPermission/IsTimeout all properly handle a `nil` error: [golang/go#31065](https://github.com/golang/go/issues/31065)

I'll refactor than and do unit test for all cases

**Comment by kopachevsky** [comment]
@dhobsd refactored code, as long it's covered by tests already, ready for re review
